extensions/redis-cache/runtime/src/main/java/io/quarkus/cache/redis/runtime/RedisCacheImpl.java

```diff
@@ -218,6 +218,34 @@ public Uni<V> apply(RedisConnection connection) {
         });
     }
 
+    @Override
+    public <K, V> V compute(K key, Class<V> clazz, BiFunction<K, V, V> command) {
+        byte[] encodedKey = marshaller.encode(computeActualKey(encodeKey(key)));
+
+        return withConnection(connection -> watch(connection, encodedKey)
+                .chain(new GetFromConnectionSupplier<>(connection, clazz, encodedKey, marshaller))
+                .chain((value) -> connection.send(Request.cmd(Command.MULTI)).replaceWith(value))
+                .chain(previousValue -> Uni.createFrom().item(() -> command.apply(key, previousValue))
+                        .runSubscriptionOn(MutinyHelper.blockingExecutor(vertx.getDelegate())))
+                .chain(value -> {
+                    if (value == null) {
+                        throw new IllegalArgumentException("Cannot cache `null` value");
+                    }
+                    byte[] encodedValue = marshaller.encode(value);
+                    return set(connection, encodedKey, encodedValue)
+                            .replaceWith(value);
+                })
+                .onFailure().call(() -> connection.send(Request.cmd(Command.DISCARD)))
+                .chain((value) -> connection.send(Request.cmd(Command.EXEC))
```

Comment:
No, you will call EXEC after DISCARD, which is illegal. 
Use a single onTermination and handle the success/failure there.